### PR TITLE
Improve the pruner for bigger programs like fabric 

### PR DESCRIPTION
**Pull request by robotrobo**

### CODE REVIEW

**Comment by fruffy** [comment]
Can you rebase your pull request? I think it might be out of sync with the master branch.

**Comment by robotrobo** [comment]
Done

File: pruner/bug.p4i [hunk 1]
```diff
@@ -0,0 +1,917 @@
+# 1 "./example_programs/fabric/bug.p4"
```
Comment (fruffy) [inline pruner/bug.p4i:1]: We  should put this in .gitignore.

File: pruner/src/statement_pruner.cpp [hunk 2]
```diff
@@ -68,14 +68,17 @@ const IR::P4Program *prune_statements(const IR::P4Program *program,
                                       uint64_t prog_size) {
     int same_before_pruning = 0;
     int result;
-    int max_statements = prog_size / SIZE_BANK_RATIO;
+    // int max_statements = prog_size / SIZE_BANK_RATIO;
```
Comment (fruffy) [inline pruner/src/statement_pruner.cpp:71]: Can you reenable this?

File: pruner/src/replace_variables.cpp [hunk 3]
```diff
@@ -42,7 +45,7 @@ const IR::P4Program *apply_replace(const IR::P4Program *program,
                               new P4::TypeInference(&refMap, &typeMap, false)});
 
     temp = program->apply(pass_manager);
-
+    emit_p4_program(temp, "before_replace_vars.p4");
```
Comment (fruffy) [inline pruner/src/replace_variables.cpp:48]: This is hard-coded, why is this necessary?

File: pruner/src/replace_variables.cpp [hunk 4]
```diff
@@ -22,6 +22,8 @@ const IR::Node *ReplaceVariables::postorder(IR::Expression *s) {
     if (typeMap->isLeftValue(expr)) {
         return s;
     }
+    // not handling NewType or structs for now 
+    if(type->is<IR::Type_Newtype>() || type->is<IR::Type_Struct>()) return s;
```
Comment (fruffy) [inline pruner/src/replace_variables.cpp:26]: Can you emit a warning here?

File: pruner/src/pruner_util.cpp [hunk 5]
```diff
@@ -262,13 +261,21 @@ int check_pruned_program(const IR::P4Program **orig_program,
...
+    int exit_code = exit_info.exit_code;
     // if got the right exit code, then modify the original program, if not
     // then choose a smaller bank of statements to remove now.
     if (exit_code != pruner_conf.exit_code) {
         INFO("FAILED");
         return EXIT_FAILURE;
     } else {
+        // When checking crash bugs, there must be the string Compiler Bug in the 
+        // output (the exit code can remain the same if there is another error)
+        if(pruner_conf.err_type == ErrorType::CrashBug && exit_info.err_msg.find("Compiler Bug") == NULL){
```
Comment (fruffy) [inline pruner/src/pruner_util.cpp:274]: CrashBug should already include this check, shouldn't it? This is how defined it. If the classification is wrong, we need to fix it earlier.

File: pruner/example_programs/fabric/fabric.p4 [hunk 6]
```diff
@@ -59,13 +108,14 @@ control FabricIngress (inout parsed_headers_t hdr,
...
     IngressQos() qos;
+
 #ifdef WITH_PORT_COUNTER
     PortCountersControl() port_counters_control;
 #endif // WITH_PORT_COUNTER
 #ifdef WITH_SPGW
     SpgwIngress() spgw;
 #endif // WITH_SPGW
-
+/*
```
Comment (fruffy) [inline pruner/example_programs/fabric/fabric.p4:118]: What is this comment doing here? Please remove it, we should keep as much from the original program as possible.

File: pruner/example_programs/fabric/fabric.p4 [hunk 7]
```diff
@@ -46,9 +46,58 @@
 #include "include/int/int_main.p4"
 #endif // WITH_INT
 
+
+//////////////////
+
+#include <pna.p4>
```
Comment (fruffy) [inline pruner/example_programs/fabric/fabric.p4:52]: This include should not be needed.

File: pruner/example_programs/fabric/fabric.p4 [hunk 8]
```diff
@@ -46,9 +46,58 @@
...
+
+
+struct headers_t {
+    ethernet_t ethernet;
+    ipv4_t ipv4;
+}
+
+struct main_metadata_t {
+   PortId_t dst_port;
+}
```
Comment (fruffy) [inline pruner/example_programs/fabric/fabric.p4:65]: Most of this should not be needed, just add a minimal function to make things compile.

**Review by fruffy — CHANGES_REQUESTED** [review]

File: pruner/src/statement_pruner.cpp (hunk 2 above)
Comment (robotrobo) [reply pruner/src/statement_pruner.cpp:71]: Alright, but we might need to play around the ratio,

File: pruner/src/pruner_util.cpp (hunk 5 above)
Comment (fruffy) [reply pruner/src/pruner_util.cpp]: We usually have the enum label Error for this and  it should be classified as you mentioned. Are we not using it?

Comment (robotrobo) [reply pruner/src/pruner_util.cpp]: Right but we only call classify_bug initially when we receive the original program, but during our pruning process we never call classify_bug, so what i'm saying is lets say we have two versions of the program before_prune_iteration1 and after_prune_iteration1, and suppose the prune iteration removed some statements from the program which caused a syntax error, so in this case we want this iteration to "fail" right instead of "pass", but in our earlier implementation we assumed that removing statements can't lead to a syntax error i think.

Comment (fruffy) [reply pruner/src/pruner_util.cpp]: We probably should reclassify every time then, to make sure that we do not lose the compiler bug. I think there was a reason why we did it differently but we should review that.

**Comment by robotrobo** [comment]
The pruner still seems to crash during the statement pruning stage, need to fix that now

**Comment by robotrobo** [comment]
The problem is that when the pruner removes something from an if block, we encounter a `Null IfTrue` compiler bug and crash.

**Comment by robotrobo** [comment]
The `NULL IFtrue` bug was caused when we prune the sole statement present in an if block, so instead we replace the if statement with a block statement with two components :- The `ifTrue` block and the `ifFalse` block.

**Comment by fruffy** [comment]
Okay so I ran this with the following command:
`../../p4c/build/p4pruner --compiler-bin ../../p4c/build/p4c-bm2-ss  --bug-type CRASH example_programs/fabric/fabric.p4 2>&1 | tee out.txt`

and got this output

```
Checking if pruned/ exists.
Ignoring validation bin for crash bug
Grabbing config by using the compiler binary.
Checking if /storage/Projekte/p4_tv/modules/p4c/build/p4c-bm2-ss exists.
Checking exit code.
Crash bug
Using generated seed.
Seed:3913899663
Size of the program :317 statements

Pruning statements
Trying with  288 statements
modules/p4c/ir/indexed_vector.h:185:58: runtime error: member access within null pointer of type 'const struct StatOrDecl'
modules/p4c/ir/indexed_vector.h:185:58: runtime error: member call on null pointer of type 'const struct Node'
Checking exit code.
FAILED
Trying with  144 statements
Checking exit code.
Crash bug
PASSED: Reduced by 19.3948 %
Trying with  146 statements
Checking exit code.
FAILED
Trying with  73 statements
Checking exit code.
FAILED
Trying with  36 statements
Checking exit code.
FAILED
Trying with  18 statements
Checking exit code.
Crash bug
PASSED: Reduced by 2.1978 %
Trying with  20 statements
Checking exit code.
Crash bug
PASSED: Reduced by 3.33507 %
Trying with  22 statements
Checking exit code.
FAILED
Trying with  11 statements
Checking exit code.
Crash bug
PASSED: Reduced by 1.46471 %
Trying with  13 statements
Checking exit code.
FAILED
Trying with  6 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.828475 %
Trying with  8 statements
Checking exit code.
FAILED
Trying with  4 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.687569 %
Trying with  6 statements
Checking exit code.
FAILED
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.349626 %
Trying with  3 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.472435 %
Trying with  5 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.237339 %
Trying with  4 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.104957 %
Trying with  3 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.245158 %
Trying with  5 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED

Pruning expressions
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
Crash bug
PASSED: Reduced by 5.53664 %
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
Crash bug
PASSED: Reduced by 0.0185832 %
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED

Reducing boolean expressions
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.

Pruning with compiler passes
Applying Generic passes...
Checking exit code.
Crash bug
PASSED: Reduced by -1.11148 %
Replacing variables...
Replacing variables with literals
Checking exit code.
Crash bug
PASSED: Reduced by 0.610294 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.307021 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.126155 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.0520117 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.0111512 %
Checking exit code.
Crash bug
PASSED: Reduced by 1.10294 %
Applying custom RemoveAllUnusedDeclarations...
Checking exit code.
FAILED
Total reduction percentage = 31.1351 %
Done. Removing ephemeral working directory.
```

Does this correlate with what you have?

**Comment by fruffy** [comment]
```
modules/p4c/ir/indexed_vector.h:185:58: runtime error: member access within null pointer of type 'const struct StatOrDecl'
modules/p4c/ir/indexed_vector.h:185:58: runtime error: member call on null pointer of type 'const struct Node'
```
Are because I use the command line option `add_cxx_compiler_option ("-fsanitize=null")` which checks for nullptr access.

File: pruner/bug.p4i (hunk 1 above)
Comment (fruffy) [reply pruner/bug.p4i:1]: Remove this program please.

File: pruner/src/statement_pruner.cpp [hunk 9]
```diff
@@ -21,6 +21,19 @@ const IR::Node *Pruner::preorder(IR::BlockStatement *s) {
     return s;
 }
 
+const IR::Node *Pruner::preorder(IR::IfStatement *s) {
+    // If ifTrue is only one statement, try to replace
+    // it with the two branches
+    if (!s->ifTrue->is<IR::BlockStatement>()) {
```
Comment (fruffy) [inline pruner/src/statement_pruner.cpp:27]: I think we can do more here, right? Remove one of the branches, or remove both. Why does this lead to a bug?

Comment (robotrobo) [reply pruner/src/statement_pruner.cpp:27]: We can surely do that, the bug is only triggered when the if block is empty, so something like this :- 

```c
    if(condition 1){

       }

   else {
          statement_1;
          .
          .
          statement_n;
    }
```

**Comment by robotrobo** [comment]
No, `../../p4c/build/p4pruner --seed 3913899663 --compiler-bin ../../p4c/build/p4c-bm2-ss --bug-type CRASH example_programs/fabric/fabric.p4 | tee out.txt` returns this - 
```
Checking if pruned/ exists.
Ignoring validation bin for crash bug
Grabbing config by using the compiler binary.
Checking if ../../p4c/build/p4c-bm2-ss exists.
Checking exit code.
Using provided seed.
Seed:3913899663
Crash bug
Size of the program :317 statements

Pruning statements
Trying with  288 statements
Checking exit code.
FAILED
Trying with  144 statements
Checking exit code.
Crash bug
PASSED: Reduced by 21.9321 %
Trying with  146 statements
Checking exit code.
FAILED
Trying with  73 statements
Checking exit code.
FAILED
Trying with  36 statements
Checking exit code.
FAILED
Trying with  18 statements
Checking exit code.
Crash bug
PASSED: Reduced by 2.5661 %
Trying with  20 statements
Checking exit code.
Crash bug
PASSED: Reduced by 3.90866 %
Trying with  22 statements
Checking exit code.
FAILED
Trying with  11 statements
Checking exit code.
Crash bug
PASSED: Reduced by 1.72687 %
Trying with  13 statements
Checking exit code.
FAILED
Trying with  6 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.979365 %
Trying with  8 statements
Checking exit code.
FAILED
Trying with  4 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.814034 %
Trying with  6 statements
Checking exit code.
FAILED
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.414461 %
Trying with  3 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.560409 %
Trying with  5 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.281784 %
Trying with  4 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.124668 %
Trying with  3 statements
Checking exit code.
Crash bug
PASSED: Reduced by 0.291254 %
Trying with  5 statements
Checking exit code.
FAILED
Trying with  2 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED
Trying with  1 statements
File has not changed. Skipping analysis.
Trying with  3 statements
Checking exit code.
FAILED

Pruning expressions
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
Crash bug
PASSED: Reduced by 6.5807 %
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
Crash bug
PASSED: Reduced by 0.0223344 %
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED
File has not changed. Skipping analysis.
Checking exit code.
FAILED

Reducing boolean expressions
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.
File has not changed. Skipping analysis.

Pruning with compiler passes
Applying Generic passes...
Checking exit code.
Crash bug
PASSED: Reduced by -1.32249 %
Replacing variables...
Replacing variables with literals
Checking exit code.
Crash bug
PASSED: Reduced by 0.731987 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.368692 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.151589 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.062514 %
Checking exit code.
Crash bug
PASSED: Reduced by 0.0134042 %
Checking exit code.
Crash bug
PASSED: Reduced by 1.32287 %
Applying custom RemoveAllUnusedDeclarations...
Checking exit code.
Crash bug
PASSED: Reduced by 75.8111 %
Total reduction percentage = 84.3297 %
Done. Removing ephemeral working directory.
```

File: pruner/src/statement_pruner.cpp (hunk 9 above)
Comment (fruffy) [reply pruner/src/statement_pruner.cpp:27]: I think this may be an assertion in the if statement. If that happens. The true part of the statement can not be nullptr. So we just can prune the else part, convert the entire if statement into a simple block statement, or replace the true part with the else part (and negate the condition)

**Comment by fruffy** [comment]
Okay I ran this with the option not to unroll includes and I got

`Total reduction percentage = 84.3297 %`

Which is quite good. 
But I see there is a lot we can still remove, for example typedefs and reduce the header size.

**Comment by fruffy** [comment]
If this is ready for review, let me know.

**Comment by robotrobo** [comment]
I'll try to have a look at the header and typedefs first, and then mark this ready.

File: pruner/src/statement_pruner.cpp (hunk 9 above)
Comment (fruffy) [reply pruner/src/statement_pruner.cpp:27]: This looks incomplete to me. Shouldn't we either delete both branches, or one or the other? Currently this always deletes the else statement.  If you delete the then part of the statement you just need to swap in the else part. If there is no else part, you can delete the entire statement. So you need a random dice roll here.

**Review by fruffy — CHANGES_REQUESTED** [review]

**Comment by robotrobo** [comment]
I was thinking if we should remove the entire statement 50% of times or 33% of times

**Comment by fruffy** [comment]
LGTM. I will clean some of this up with the other pull requestion. Great stuff!

We should also make all the percentages global configurable constants. I will take care of that.

**Review by fruffy — APPROVED** [review]
